### DISCUSSION TITLE
Setting media

### DISCUSSION THREAD

**Discussion by sophiez8** [micom-dev/micom#90] [Q&A] [answered]
Dear Christian,

thank you for answering my previous questions! 

Referring to query #88 in the QnA I have some remaining issues when setting the media and therefore have some remaining questions.
I can indeed see that MICOM created a single media exchange reaction for a metabolite from both models, when I ran FBA without constraints/media specifications. From those FBA results I have obtained the ‘media exchange reaction names’ that end in “…._m” . 

1.	However, when I try to set the media upper bound fluxes I get the error that ‘No ID from the medium could be found in the exchange reactions’ when I set the media:  

import pandas as pd
data = [['EX_thymine_m',10],['EX_3_5_dimethoxytoluene_m',10], ……… , ['EX_s_4262_m',0]]
med = pd.DataFrame(data,columns=['reaction','flux'], dtype=float)
com.medium = med

ValueError                                Traceback (most recent call last)
/var/folders/w8/tl2htjns5zx56ncq5plltjg00000gr/T/ipykernel_8325/420956091.py in <module>
      6 #med
      7 #medium
----> 8 com.medium = med
      9 
     10 

~/opt/anaconda3/lib/python3.9/site-packages/micom/community.py in medium(self, fluxes)
    662         not_found = rids - exids
    663         if len(found) == 0:
--> 664             raise ValueError(
    665                 "No ID from the medium could be found in the exchange reactions. "
    666                 "This means you probably have mismatched IDs..."

ValueError: No ID from the medium could be found in the exchange reactions. This means you probably have mismatched IDs...

I’m confused to get this error since I have set the media IDs according to the ones I have obtained previously from the FBA output, where I did not set any media. 

2.	When I try to simulate growth under that medium I get the Error ‘Check that you have CPLEX or Gurobi installed’, however micom automatically downloads OSQP as a solver so there should be no need for CPLEX or Gurobi?  

import pandas as pd
from micom.workflows import tradeoff
data = [['EX_thymine_m',1000],['EX_3_5_dimethoxytoluene_m',1000], ……. ['EX_s_4262_m',0]]
med = pd.DataFrame(data,columns=['reaction','flux'], dtype=float)
from micom.workflows import grow
res = grow(manifest, model_folder="Downloads", medium=med, tradeoff= None, threads=1, atol = None, rtol=None, presolve=True)

[18:19:11] ERROR    Could not solve cooperative tradeoff for sample_1. This can    [grow.py](https://github.com/Users/Sophie3/opt/anaconda3/lib/python3.9/site-packages/micom/workflows/grow.py):[63](https://github.com/Users/Sophie3/opt/anaconda3/lib/python3.9/site-packages/micom/workflows/grow.py#63)
                    often be fixed by enabling `presolve`, choosing more                     
                    permissive atol and rtol arguments, or by checking that medium           
                    fluxes are > atol.                                                       
---------------------------------------------------------------------------
OptimizationError                         Traceback (most recent call last)
/var/folders/w8/tl2htjns5zx56ncq5plltjg00000gr/T/ipykernel_8325/2749939476.py in <module>
      6 from micom.workflows import grow
      7 
----> 8 res = grow(manifest, model_folder="Downloads", medium=med, tradeoff= None, threads=1, atol = None, rtol=None, presolve=True)
      9 
     10 #from micom.workflows import tradeoff

~/opt/anaconda3/lib/python3.9/site-packages/micom/workflows/grow.py in grow(manifest, model_folder, medium, tradeoff, threads, weights, strategy, atol, rtol, presolve)
    188     results = workflow(_growth, args, threads)
    189     if all([r is None for r in results]):
--> 190         raise OptimizationError(
    191             "All numerical optimizations failed. This indicates a problem "
    192             "with the solver or numerical instabilities. Check that you have "

OptimizationError: All numerical optimizations failed. This indicates a problem with the solver or numerical instabilities. Check that you have CPLEX or Gurobi installed. You may also increase the abundance cutoff to create simpler models.

Could you please further explain what “choosing more permissive atol and rtol arguments, or by checking that medium fluxes are > atol “ means? I have also gotten the same error before setting atol and rtol to 0.                                                        

This may be relevant: I get an Attribute Error for the following: 

from micom.workflows import tradeoff
tradeoff_rates = tradeoff(manifest,model_folder="Downloads", medium=medium, threads=2)
tradeoff_rates.head()

AttributeError: 'NoneType' object has no attribute 'members'

3.	Is it possible that these errors arise because the two models have differently defined directionality for uptakes and would this present an issue? Can this issue of different directionalities be overcome in micom somehow?

Kind regards,

Sophie

**Comment by cdiener** [comment, marked as answer]
For setting `com.medium` the passed medium must be dict-like object like in cobrapy. Media in DataFrame object are only accepted in the `grow` workflow. So in your example it would be:

```python
...
med.index = med.reaction
com.medium = med.flux
...
```

**Reply by sophiez8** [reply]
thank you!
